File: csharp/2015/day03a/Program.cs
using System;
using System.Collections.Generic;
using System.IO;

using AOC_Utils.Geometry;

class Program
{
    static void Usage()
    {
        Console.Error.WriteLine("usage: dotnet run <input file>");
        Environment.Exit(1);
    }

    static int Process(string filename)
    {
        HashSet<Position2D<int>> positions = new HashSet<Position2D<int>>();

        Position2D<int> santa = new Position2D<int>(0 , 0);
        positions.Add(santa);

        try
        {
            using ( StreamReader reader = new StreamReader(filename) )
            {
                string? line;
                while ( (line = reader.ReadLine()) != null )
                {
                    foreach ( char ch in line )
                    {
                        switch ( ch )
                        {
                            case '^':
                                santa.Y += 1;
                                break;
                            case 'v':
                                santa.Y -= 1;
                                break;
                            case '<':
                                santa.X -= 1;
                                break;
                            case '>':
                                santa.X += 1;
                                break;
                        }
                        positions.Add(santa);
                    }
                }
            }
        }
        catch ( IOException e )
        {
            Console.Error.WriteLine("Error reading file: " + e.Message);
            Environment.Exit(1);
        }

        return positions.Count;
    }

    static void Main(string[] args)
    {
        if ( args.Length < 1 )
        {
            Usage();
        }

        string filename = args[0];
        int result = Process(filename);
        Console.WriteLine("result = " + result);
    }
}

File: fsharp/2015/day03a/Program.fs
open System
open System.Collections.Generic
open System.IO

open AOC_Utils.Geometry

let usage() =
    Console.Error.WriteLine("usage: dotnet run <input file>")
    Environment.Exit(1)

let processFile(filename: string) =
    let mutable positions = HashSet<Position2D<int>>()
    let f (santa: Position2D<int>) ch =
        let santa' = match ch with
                     | '^' -> new Position2D<int>(santa.X, santa.Y + 1)
                     | 'v' -> new Position2D<int>(santa.X, santa.Y - 1)
                     | '<' -> new Position2D<int>(santa.X - 1, santa.Y)
                     | '>' -> new Position2D<int>(santa.X + 1, santa.Y)
                     | _ -> santa
        positions.Add(santa') |> ignore
        santa'

    try
        File.ReadAllText(filename)
        |> Seq.fold f (Position2D<int>(0, 0))
        |> ignore
        positions.Count
    with
    | :? IOException as e ->
        Console.Error.WriteLine("Error reading file: " + e.Message)
        Environment.Exit(1)
        0

[<EntryPoint>]
let main args =
    if args.Length < 1 then
        usage()

    let filename = args.[0]
    let result = processFile filename
    Console.WriteLine("result = " + result.ToString());
    0 
